### DISCUSSION TITLE
How to correctly perform jax.lax.scan over multiple PyTree?

### DISCUSSION THREAD

**Discussion by yura-hb** [jax-ml/jax#13451] [Q&A] [answered]
Hi, I want to perform `jax.lax.scan` over multiple PyTrees. The example of what I'm trying to achieve is the following. The problem is that the `scanned` function is called with the whole `xs` sequence instead of being called with tuple instead of the tuple (jnp.array([1]), jnp.array([2]), jnp.array([4])).
I suppose that it should be possible to achieve this because python list is pytree object. However, I'm not sure, how can I make `jax` treat each tuple as the leaf of the tree.
 
```
import jax.numpy as jnp

# arrays here are the example, actually I have multiple lists of pytree objects, which I can't stack into `jnp.ndarray`
a = [jnp.array([1]), jnp.array([2]), jnp.array([3])]
b = [jnp.array([2]), jnp.array([3]), jnp.array([4])]
c = [jnp.array([4]), jnp.array([5]), jnp.array([6])]

xs = list(zip(a, b, c))

def scanned(accumulator, pair):
    _a, _b, _c = pair

    print(len(_a), len(_b), len(_c)) # Outputs (3, 3, 3)

    return accumulator, 0

print(jax.lax.scan(scanned, init=10, xs=xs))

```

**Comment by jakevdp** [comment, marked as answer]
`scan` only operates over arrays, not over lists, so there is no way to scan over a list of pytrees.

If you want to use `scan` over pytrees, you'll have to re-express your computation using a *struct-of-arrays* pattern, rather than an *array-of-structs* pattern: that is, the input must be a single pytree (here a list) whose elements are JAX arrays.

From the example you give, it's difficult for me to tell what the intent was (are you hoping for `scan` to iterate over the outer or inner lists?) but you can modify your data like this to make it a struct of arrays, and your function will execute without error:
```python
xs = [jnp.array(x) for x in xs]
```

**Reply by yura-hb** [reply]
> Are you hoping for scan to iterate over the outer or inner lists?

I'm hoping to iterate over outer lists. So in the first iteration, I want to receive (a[0], b[0], c[0]).


The idea is to rewrite this part of the code with `lax.scan`. It is simple backward pass of the layered nn. `_state` and `_grad_state` are instances of `NamedTuple` and `layer` is the registered pytree node class.

```
result = gradient
new_states = []

for layer, _state, _grad_state in zip(reversed(self.layers), reversed(state.states), reversed(grad_state.states)):
    result, new_state = function(layer, result, _state, _grad_state)
    new_states += [new_state]

return result, State(list(reversed(new_states)))
```

> the input must be a single pytree (here a list) whose elements are JAX arrays.

This approach also wouldn't work (Modified to better show use case).

```
class SOA(NamedTuple):
    a: List[jnp.ndarray]
    b: List[jnp.ndarray]
    c: List[jnp.ndarray]

class A(NamedTuple):
    value: jnp.ndarray

class B(NamedTuple):
    value: jnp.ndarray

class C(NamedTuple):
    value: jnp.ndarray

a = [A(jnp.array([1])), A(jnp.array([2])), A(jnp.array([3]))]
b = [B(jnp.array([2.])), B(jnp.array([3.])), B(jnp.array([4.]))]
c = [C(jnp.array([4])), C(jnp.array([5])), C(jnp.array([6]))]

def scanned(accumulator, pair):
    _a, _b, _c = pair

    print(_a, _b, _c)

    return accumulator, 0

jax.lax.scan(scanned, init=10, xs=SOA(a, b, c))
jax.lax.scan(scanned, init=10, xs=(a, b, c)) # Nor this
```

Following that, It seems, that with lax it isn't possible to preserve the identity of type. The instance tree must be manually flattened and then unflattened?


Something like this can work, if there is no better way(.

```
a = [A(jnp.array([1])), A(jnp.array([2])), A(jnp.array([3]))]
b = [B(jnp.array([2.])), B(jnp.array([3.])), B(jnp.array([4.]))]
c = [C(jnp.array([4])), C(jnp.array([5])), C(jnp.array([6]))]

xs = [jnp.array(x) for x in (a, b, c)]

def scanned(accumulator, pair):
    _a, _b, _c = pair

    a = A(*_a).value
    b = B(*_b).value
    c = C(*_c).value

    return accumulator, (a + b + c)

print(jax.lax.scan(scanned, init=10, xs=xs)[1])
```

**Reply by jakevdp** [reply]
It's not that the instance must be manually flattened and unflattened; rather it's that JAX scan operates on structs of arrays rather than arrays of structs. If you want to use jax.scan, you'll have to change how you represent your layers to fit a struct-of-arrays pattern.

**Reply by yura-hb** [reply]
Ok, thanks
